lib_i2s/src/i2s_frame_master_impl.h

```diff
@@ -40,103 +49,220 @@ static i2s_restart_t i2s_frame_ratio_n(client i2s_frame_callback_if i2s_i,
+                            :"r"(p_din[i])
+                            :"memory");
+                in_samps[idx] = bitrev(data);
+            }
+
+            lr_mask = ~lr_mask;
+            p_lrclk <: lr_mask;
         }
+        else
+        {
+#pragma loop unroll
+            for (size_t i = 0, idx = 0; i < num_in; i++, idx += 2)
+            {
+                int32_t data;
+                asm volatile("in %0, res[%1]"
+                            :"=r"(data)
+                            :"r"(p_din[i])
+                            :"memory");
+                set_port_shift_count(p_din[i], num_data_bits);
+                in_samps[idx] = bitrev(data) & data_bit_mask;
+            }
 
-        lr_mask = ~lr_mask;
-        p_lrclk <: lr_mask;
+            lr_mask = ~lr_mask;
+            partout(p_lrclk, num_data_bits, lr_mask);
+        }
 
-        if (restart == I2S_NO_RESTART) {
-            //Output i2s odds (1,3,5..)
+        if (restart == I2S_NO_RESTART) 
+        {
+            // Output i2s odds (1,3,5..)
+            if (num_data_bits == 32)
+            {
 #pragma loop unroll
-            for (size_t i=0, idx=1; i<num_out; i++, idx+=2){
-                p_dout[i] <: bitrev(out_samps[idx]);
+                for (size_t i = 0, idx = 1; i < num_out; i++, idx += 2)
+                {
+                    p_dout[i] <: bitrev(out_samps[idx]);
+                }
+
+                lr_mask = ~lr_mask;
+                p_lrclk <: lr_mask;
             }
+            else
+            {
+#pragma loop unroll
+                for (size_t i = 0, idx = 1; i < num_out; i++, idx += 2)
+                {
+                    partout(p_dout[i], num_data_bits, bitrev(out_samps[idx] << data_bit_offset));
+                }
 
-            lr_mask = ~lr_mask;
-            p_lrclk <: lr_mask;
+                lr_mask = ~lr_mask;
+                partout(p_lrclk, num_data_bits, lr_mask);
+            }
         }
 
-        //Input i2s odds (1,3,5..)
+        // Input i2s odds (1,3,5..)
+        if (num_data_bits == 32)
+        {
 #pragma loop unroll
-        for (size_t i=0, idx=1; i<num_in; i++, idx+=2){
-            int32_t data;
-            asm volatile("in %0, res[%1]":"=r"(data):"r"(p_din[i]):"memory");
-            in_samps[idx] = bitrev(data);
+            for (size_t i = 0, idx = 1; i < num_in; i++, idx += 2)
+            {
+                int32_t data;
+                asm volatile("in %0, res[%1]"
+                            :"=r"(data)
+                            :"r"(p_din[i])
+                            :"memory");
+                in_samps[idx] = bitrev(data);
+            }
         }
+        else
+        {
+#pragma loop unroll
+            for (size_t i = 0, idx = 1; i < num_in; i++, idx += 2)
+            {
+                int32_t data;
+                asm volatile("in %0, res[%1]"
+                            :"=r"(data)
+                            :"r"(p_din[i])
+                            :"memory");
```

Comment:
Don't think this memory clobber is required?

Reply:
You're very right - have removed it (and all others of its form) in a later commit.